### DISCUSSION TITLE
Using SQLAlchemy 2 Mapped Classes in combination with PyCharm

### DISCUSSION THREAD

**Discussion by teuneboon** [sqlalchemy/sqlalchemy#9504] [Usage Questions]
For a project we use SQLAlchemy 1.4 with async support, since the migration to SQLAlchemy 2 was relatively simple we decided to go through with it since we're pre-release anyway. The main part of our migration was migrating from the old `Column()` syntax to the new Mapped Classes. Example I made to reproduce the issue:

**OLD**
```python
from sqlalchemy import String, Column, Integer
from sqlalchemy.orm import declarative_base


Base = declarative_base()


class Car(Base):
    __tablename__ = "cars"
    id: int = Column(Integer, primary_key=True)
    model: str = Column(String(80))

    def testing(self):
        self.model = "Nissan"
```

**NEW**
```python
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import mapped_column, Mapped


class Base(DeclarativeBase):
    pass


class Car(Base):
    __tablename__ = "cars"
    id: Mapped[int] = mapped_column(primary_key=True)
    model: Mapped[str] = mapped_column(String(80))

    def testing(self):
        self.model = "Nissan"
```

The issue lies in the `self.model = "Nissan"` line, in the old code PyCharm was perfectly happy with that line of code, but after migrating it has the following issue with it: 
![image](https://user-images.githubusercontent.com/571486/225941005-229d651b-2bd9-4829-8741-97e459161228.png)

Does anyone have some tips and/or solutions on how to deal with this? I couldn't really find an open issue about this on either SQLAlchemy's or PyCharm's issue trackers, but I'm not quite sure what to search for so I might've missed something, and I don't want to just disable these type of warnings straight out.

**Comment by zzzeek** [comment]
that would be a bug in Pycharm, unfortunately.   `Mapped` is a [Python descriptor](https://docs.python.org/3/howto/descriptor.html) which indicates that the "self" -level datatype for ``model`` is ``str``.

the code you have, if we add `-> None` return type for your `def testing()` passes in pylance (vscode)/pyright as well as Mypy.  to eliminate any issues with stubs, make sure mypy and/or pyright shows no errors with your code.

**Reply by teuneboon** [reply]
I'll see if I can reproduce the issue to report to PyCharm without telling them to just use the whole SQLAlchemy library(simple reproduction cases often do better).

For the mypy part: interestingly enough mypy doesn't seem to consider the lacking `-> None` to be an issue here, in fact, it seems to consider the following correct:

```python
from sqlalchemy import String
from sqlalchemy.orm import Mapped, DeclarativeBase, mapped_column


class Base(DeclarativeBase):
    pass


class Car(Base):
    __tablename__ = "cars"
    id: Mapped[int] = mapped_column(primary_key=True)
    model: Mapped[str] = mapped_column(String(80))

    def testing(self):
        self.model = "Nissan"
        self.model = 123  # <-- this should be super incorrect


if __name__ == '__main__':
    car = Car()
    car.testing()
    print(car.model)
```

```
> ~/.pyenv/versions/3.11.2/bin/python -m mypy testing.py
Success: no issues found in 1 source file
> ~/.pyenv/versions/3.11.2/bin/python -m mypy --version
mypy 1.1.1 (compiled: yes)
```

(when I try to do `car = '123'` at the end it does correctly give an error, so it doesn't seem like my whole mypy is broken somehow)

Pyright does see the `self.model = 123` line as an issue however

**Comment by Gloweye** [comment]
Hi, @teuneboon 's colleague here. I was the one to initially run into this issue. 

It can be resolved by adding the following right before the `Car` class:

```python
if TYPE_CHECKING:
    
    T = TypeVar('T')
    
    
    class Mapped(Generic[T]):
        def __set_name__(self, owner: type[Base], name: str) -> None:
            self.name = name

        @overload
        def __get__(self, obj: None, objtype: None) -> 'Mapped':
            ...

        def __get__(self, obj: Base, objtype: type[Base]) -> T:
            ...

        def __set__(self, obj: Base, value: T) -> None:
            ...

```

We could of course shadow our own Mapped this way, but that doesn't sound all that neat. 

Following the declaration to [sqlalchemy/orm/base.py](../blob/main/lib/sqlalchemy/orm/base.py#L711), we find the type hints. However, there is no type-hint for the `__set__` method at all, which I believe to be the reason Pycharm complains - it thinks we're trying to replace the descriptor (type: Mapped), with the value (type: int)

**Reply by CaselIT** [reply]
Mapped has set declared https://github.com/sqlalchemy/sqlalchemy/blob/fa0666cb04174cdd2592ab1327d48e431fe86ffa/lib/sqlalchemy/orm/base.py#L775

**Reply by Gloweye** [reply]
Huh, indeed. Makes me curious why it likes my redefinition more.

**Reply by zzzeek** [reply]
have you posted a bug report w/ pycharm?

**Reply by teuneboon** [reply]
Sorry for the late response, just adding the PyCharm issue to this discussion so people landing on this discussion in the future know where to look: https://youtrack.jetbrains.com/issue/PY-59896
